### DISCUSSION TITLE
Staged files automatically added to stashes

### DISCUSSION THREAD

**Discussion by NoahLerner** [gitextensions/gitextensions#13315] [Q&A]
## Current behaviour

When executing a partial stash, files that are staged (under the "commit" dialog) are automatically included in the stash.

## Expected behaviour

Only selected files in the "stash" dialog should be stashed regardless if there are staged files for commit.

## Steps to reproduce

1. Stage a file for commit
2. Go to Manage Stashes
3. Stash a different file using partial stash and custom stash name
4. Open the saved stash and see that the staged file also appears

Could be that the bug appears in full stash, or without a custom message, or any combination. Above is one scenario to reproduce.

## Did this work in previous version of GitExtensions

Don't know, sorry.

## Environment

- Git Extensions 3.0.1.4933
- Build a269358b22a5dc1ea36928bb87f9e535de76a058
- Git 2.20.1.windows.1
- Microsoft Windows NT 10.0.15063.0
- .NET Framework 4.7.3260.0
- DPI 120dpi (125% scaling)

**Comment by gerhardol** [comment]
This ishow _git stash push_ works, index is always included.
No possibility to exclude them what I know.

The UI could mark index files in some special way to point this out though.

**Comment by NoahLerner** [comment]
Thanks for clearing that up.

**Comment by montoner0** [comment]
I second about making changes to the UI so staged files will be clearly visible as stashed.

**Comment by SAPHemlock** [comment]
Checked this against master and real git, and it looks like working-as-designed rather than something Git Extensions adds - plus the UI improvement asked for in the comments has since landed.

**What GE actually runs.** `Commands.StashSave` builds exactly `git stash push -m "msg" -- <selected paths>` for a partial stash; nothing extra is appended. The staged file is included because of how git builds the stash commit - the worktree commit is made *on top of the index*, so index content is part of the stash tree.

Scratch repo on git 2.55, `A.txt` staged, `B.txt` unstaged, stashing only `B.txt`:

```
$ git stash push -m msg -- B.txt
$ git status --short
M  A.txt          <- still staged, nothing was taken away
 M C.txt
$ git diff --stat stash@{0}^ stash@{0}
 A.txt | 1 +      <- but the staged change is also inside the snapshot
 B.txt | 1 +
```

There is no git option that excludes the index from a stash. I checked both candidates:

- `--keep-index` does not do it - it only restores the index into the worktree afterwards; the stash tree still contains both files.
- `--staged` (git 2.35+) is the opposite, it stashes *only* the index, and it rejects a pathspec outright (`No staged changes`).

That matches @gerhardol's comment above.

**The follow-up ask is done.** The suggestion to "mark index files in some special way" was implemented in `d6b5d14e7` "FormStash: Diff for index files were not displayed - Show worktree/index separatly": `FormStash` now splits the list and renders two labelled groups, Index and Workspace, so when picking files for a partial stash you can see which ones are staged.

```csharp
List<GitItemStatus> indexItems = [.. gitItemStatuses.Where(item => item.Staged == StagedStatus.Index)];
List<GitItemStatus> workTreeItems = [.. gitItemStatuses.Where(item => item.Staged != StagedStatus.Index)];
Stashed.SetStashDiffs(headRev, indexRev, TranslatedStrings.Index, indexItems, workTreeRev, TranslatedStrings.Workspace, workTreeItems);
```

"Keep index" and "Include untracked files" are both on the stash dialog, and "Stash staged changes" exists separately for git >= 2.35.

So this can probably be closed as by-design. Implementing true exclusion would mean hand-rolling stash creation with plumbing (temp index from HEAD, apply selected worktree changes, `write-tree`/`commit-tree`, force `refs/stash`, then checkout paths), which is a lot of risk around uncommitted work for no gain over what git offers.

One genuinely remaining gap, if anyone wants a follow-up issue: when *viewing an existing* stash, `FormStash` falls back to a flat diff over `stash^..stash` and `GetStashDiffFiles` only consults `stash^` and `stash^3` (untracked), never `stash^2` (the index commit). So a saved stash still mixes index-origin and worktree-origin files without the Index/Workspace split that the pre-stash view now has.

**Comment by gerhardol** [comment]
> One genuinely remaining gap, if anyone wants a follow-up issue: when viewing an existing stash, FormStash falls back to a flat diff over stash^..stash and GetStashDiffFiles only consults stash^ and stash^3 (untracked), never stash^2 (the index commit). So a saved stash still mixes index-origin and worktree-origin files without the Index/Workspace split that the pre-stash view now has.

if visible in the grid, index is shown too. From what I have heard, users are confused about that.
But it would probably be slightly more correct to show index also for saved stashes.
